Fixes #2865. Unknown character sequence while resizing terminal.

BDisp opened the pull request:
Fixes #2865 - On macOS Ansi escape sequences are send after the terminal resizing but they are corrupted. Thus is needed to restore them correctly. Also this fix includes some keys that aren't handled on macOS if a `Wiindows` numeric keyboard is used, like the Home, End, PgUp, PgDn, etc.... Another issue is that macOS doesn't rely on the curses `lines_ptr` and `cols_ptr` pointers to return the current size on resizing and now it's needed to use `ioctl` and `resizeterm` to get and set the current size. The better way to wait a Esc or an Ansi escape sequence, is use the `nodelay` method. I also added a new scenario to test the `Terminal.Gui` doesn't freeze while resizing.

## Pull Request checklist:

- [x] I've named my PR in the form of "Fixes #issue. Terse description."
- [x] My code follows the [style guidelines of Terminal.Gui](https://github.com/gui-cs/Terminal.Gui/blob/develop/.editorconfig) - if you use Visual Studio, hit `CTRL-K-D` to automatically reformat your files before committing.
- [x] My code follows the [Terminal.Gui library design guidelines](https://github.com/gui-cs/Terminal.Gui/blob/develop/CONTRIBUTING.md)
- [x] I ran `dotnet test` before commit
- [ ] I have made corresponding changes to the API documentation (using `///` style comments)
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any poor grammar or misspellings
- [x] I conducted basic QA to assure all features are working

kacperpikacz commented:
When running on MacBook M2 the application crashes immediately.

When I revert changes in`console_sharp_get_dims (out int lines, out int cols)` the problem disappears and I see no more issues.


<img width="1365" alt="Screenshot 2023-11-01 at 11 52 28 PM" src="https://github.com/gui-cs/Terminal.Gui/assets/76921244/c4291503-2bdb-43ff-9ecb-709189d59d47">

BDisp commented:
Thanks. The problem is when I resize the terminal the app doesn't resize. It seem that some method is missing. Can you find what? Maybe ioctl or resizeterm.

BDisp commented:
I added a definition for toggle the use of `ioctl` and now it's using the previous behavior. I now see that the issue on resizing isn't happening anymore. Maybe was other code that was causing the issue but I already fixed. Please confirm if it's allright now. Thanks.

kacperpikacz commented:
I can't make it work with c#
```
using System.Runtime.InteropServices;

namespace Playground {
	class Program {
		[DllImport ("libc", SetLastError = true)]
		internal static extern int ioctl (int fd, ulong cmd, out winsize argp);

		public static ulong TIOCGWINSZ = 0x40087468;

		public struct winsize {
			public ushort ws_row;
			public ushort ws_col;
			public ushort ws_xpixel;
			public ushort ws_ypixel;
		}

		static void Main (string [] args)
		{
			var result = ioctl (1, TIOCGWINSZ, out winsize ws);

			Console.WriteLine (result);
			if (result == 0) {
				Console.WriteLine ($"Rows: {ws.ws_xpixel}, Columns: {ws.ws_ypixel}");

			} else {
				Console.WriteLine ($"Error code: {Marshal.GetLastWin32Error ()}");
			}

			Console.ReadKey ();
		}
	}
}

```
https://github.com/gui-cs/Terminal.Gui/assets/76921244/db1c00a9-36ad-4357-87d9-e9ee095a8567

But C++ will always print correct size.
```
#include <iostream>
#include <sys/ioctl.h>
#include <unistd.h>

int main() {
    struct winsize w;
    if (ioctl(STDOUT_FILENO, TIOCGWINSZ, &w) == 0) {
        
        
        std::cout << STDOUT_FILENO << std::endl;
        std::cout << TIOCGWINSZ << std::endl;
        std::cout << "Terminal size: " << w.ws_row << " rows x " << w.ws_col << " columns" << std::endl;
    } else {
        std::cerr << "Failed to get terminal size." << std::endl;
        return 1;
    }

    return 0;
}
```

kacperpikacz commented:
> I added a definition for toggle the use of `ioctl` and now it's using the previous behavior. I now see that the issue on resizing isn't happening anymore. Maybe was other code that was causing the issue but I already fixed. Please confirm if it's allright now. Thanks.

Your last commit fixes the problem. No more issues.

BDisp commented:
Try with this value `TIOCGWINSZ = 0x5413` on c# please. Thanks.

kacperpikacz commented:
> Try with this value `TIOCGWINSZ = 0x5413` on c# please. Thanks.
> 
> Edit: If that not works, please see what value `TIOCGWINSZ` has on C++ please.

Now ioctl returns -1, and win32error is 0.
C++ using 0x40087468

BDisp commented:
> Now ioctl returns -1, and win32error is 0.
> C++ using 0x40087468

Do you know why you're having the issue before?

BDisp commented:
> Now ioctl returns -1, and win32error is 0.

In this situation you still have used `TIOCGWINSZ = 0x40087468` right?

tig commented:
@BDisp  you are chasing your tail here. 

Please see my fix to #2610. You'll note that the `Chinese.cs` scenario works fine:

- Conhost
  - NetDriver 
  - ANSIDriver

- WT
  - NetDriver
  - ANSIDriver  

Please stop trying to hack WindowsDriver to fix this; it's not worth it IMO. If people want to use conhost, they should use netdriver. 

Note `ANSIDriver` is a LONG way from being finished, but it is the right long term architecture. I plan on making it the basis for an updated curses driver too!

kacperpikacz commented:
> > Now ioctl returns -1, and win32error is 0.
> > C++ using 0x40087468
> 
> Do you know why you're having the error before?

Before https://github.com/gui-cs/Terminal.Gui/pull/2951/commits/2411b8a32e2c8b3895792dee5c7bc522ff608d04 ioctl was enabled but always return empty winsize and somehow it leaded to EntryPointNotFoundException.

BDisp commented:
> > > Now ioctl returns -1, and win32error is 0.
> > > C++ using 0x40087468
> > 
> > 
> > Do you know why you're having the error before?
> 
> Before [2411b8a](https://github.com/gui-cs/Terminal.Gui/commit/2411b8a32e2c8b3895792dee5c7bc522ff608d04) ioctl was enabled but always return empty winsize and somehow it leaded to EntryPointNotFoundException.
> 
> The correct `TIOCGWINSZ` for my Mac M2 is `0x4008746` (c++ code print this value with correct terminal size)

Now I'm confused. You said before `C++ using 0x40087468` and now you are saying it return `0x4008746`. Did you forgot to include the `8` at final?

kacperpikacz commented:
@BDisp 
<img width="823" alt="Screenshot 2023-11-03 at 1 51 47 PM" src="https://github.com/gui-cs/Terminal.Gui/assets/76921244/5ceff0d3-bf0c-4626-bb89-aa6c3106976e">

BDisp commented on Terminal.Gui/ConsoleDrivers/CursesDriver/CursesDriver.cs:
```diff
@@ -425,6 +442,56 @@ void ProcessInput ()
+								code = Curses.get_wch (out wch2);
+								if (code == 0) {
+									switch (wch2) {
+									case 16:
```
Shift.

tig replied:
Sorry, I wasn't clear. 

I was asking to re-enforce my ask above "Please add and use constants for numerics like this. Readers should not need to remember/guess what these integers mean."

BDisp replied:
The code `keyModifiers.Shift = true;` already is intuitive to inform it's a `Shift` key code.

tig replied:
I am not the only programmer in the world that struggles to remember what numbers mean. 

I do not remember that "16" means shift. I should not have to parse the code to figure that out.

BDisp replied:
@tig on macOS when I press `Shift` I get the code 16. I don't know if it's documented.

BDisp commented on Terminal.Gui/ConsoleDrivers/CursesDriver/CursesDriver.cs:
```diff
@@ -379,6 +380,24 @@ void ProcessInput ()
 				} else if (wch >= 325 && wch <= 327) { // Shift+Alt+(F1 - F3)
 					wch -= 60;
 					k = Key.ShiftMask | Key.AltMask | MapCursesKey (wch);
+				} else {
+					code = Curses.get_wch (out wch);
+					if (code == 0) {
+						switch (wch) {
+						case 16:
+							keyModifiers.Shift = true;
+							break;
+						default:
+							if (lastWch == Curses.KeyResize && wch == 91) {
+								// Returns this keys to the std input
+								Curses.ungetch (91);
+								Curses.ungetch (27);
```
It will result on CSI (`\x1b[`).

BDisp replied:
Done.
